drv/lpc55-swd/src/main.rs

```diff
@@ -1748,195 +1725,151 @@ impl ServerImpl {
 
         if self.do_setup_swd().is_ok() {
             ringbuf_entry!(Trace::SetupSwdOk);
-            need_undo |= Undo::SWD;
         } else {
             // We may have interrupted dumper or watchdog activity.
             return false; // Cannot make the required measurement.
         }
 
-        // Armv7-M Arch Ref:
-        // C1.4.1 Entering Debug state on leaving reset state
-        //
-        // To force the processor to enter Debug state as soon as it
-        // comes out of reset, a debugger sets DHCSR.C_DEBUGEN to 1, to
-        // enable Halting debug, and sets DEMCR.VC_CORERESET to 1 to
-        // enable vector catch on the Reset exception. When the
-        // processor comes out of reset it sets DHCSR.C_HALT to 1,
-        // and enters Debug state.
-
-        // If we are late to the SP_RESET party, we're still not that late.
-        // In any case, keep/force the SP into a reset condition.
-        // Though AIRCR::SYSRESETREQ can be used to effect a local reset.
-        // that does not affect the whole SP SoC.
-        // So, use the SP_RESET GPIO.
-        //
-
-        // Setting up to inject the measurement program into the SP
-        // has several potential failures. Use this `prep` closure
-        // and `need_undo` state to keep from indenting too much.
-        let mut prep = || -> Result<(), ()> {
-            self.sp_reset_enter();
-            need_undo |= Undo::RESET;
-
-            // Asserting SP_RESET for >1ms here works.
-            hl::sleep_for(1);
-
-            // Try to undo the change in DEBUGEN even if
-            // setting it failed.
-            need_undo |= Undo::DEBUGEN;
-            if self.dp_write_bitflags::<Dhcsr>(Dhcsr::resume()).is_err() {
-                ringbuf_entry!(Trace::DemcrWriteError);
-                return Err(());
-            }
+        let out = self.reset_and_measure_sp();
+        self.swd_finish();
+        out.is_ok()
+    }
+
+    /// Resets the SP, forcing it to come up in debug halted state
+    ///
+    /// This is done per C1.4.1 in the ARMv7-M Architectural Reference Manual:
+    /// - `DHCSR.C_DEBUGEN` enables halting debug
+    /// - `DEMCR.VC_CORERESET` enables vector catch on the reset exception
+    ///
+    /// When the reset pin is released, we wait for a hard-coded timeout for
+    /// `DFSR.HALTED` to be set to 1.
+    ///
+    /// If this function returns `Ok(())`, then the SP should be in debug halt;
+    /// `DHCSR.C_DEBUGEN` and `DFSR.HALTED` should be set.
+    ///
+    /// If it returns an error, then the `DHCSR.C_DEBUGEN` bit is cleared, and
+    /// the SP should be assumed to be running.
+    ///
+    /// In all cases, the reset line is not asserted when this function returns.
+    /// Additionally, `DEMCR.VC_CORERESET` is always cleared, so future resets
+    /// will not trigger a vector catch.
+    fn reset_into_debug_halt(&mut self) -> Result<(), ()> {
```

Comment:
I believe there's [a clippy lint about using `()` as an error type](https://rust-lang.github.io/rust-clippy/master/index.html#result_unit_err), which we should either allow here or turn off globally[^1]

[^1]: Personally, I don't _love_ this lint. Especially in embedded projects where a meaningfully formattable error type might have some binary size cost and won't get formatted to a user anyway, it's just saying "go make a zero sized struct for your thing", which I think is often not actually worth the effort...

Reply:
Done in f47afa139ff0